Avoid leaving connection in broken state after INSERT finishes with exception

azat opened the pull request:
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Avoid leaving connection in broken state after INSERT finishes with exception

In case of exception during processing INSERT, it is possible to leave in the state when the will be pending Data packets, and you will get something like this in logs:

    2025.01.17 20:32:31.228897 [ 1956 ] {5e4cd57a-7f69-4a7d-bc02-5948f528bebc} <Debug> executeQuery: (from [::ffff:127.0.0.1]:56486) insert into in_dist from infile 'test.tsv' settings min_insert_block_size_rows=0, min_insert_block_size_bytes=0 (stage: Complete)
    2025.01.17 20:32:31.248667 [ 1956 ] {5e4cd57a-7f69-4a7d-bc02-5948f528bebc} <Error> executeQuery: Code: 48. DB::Exception: Method write is not supported by storage SystemOne: while pushing to view default.mv (3e957e5d-8df3-4b84-b940-2aa5d4ad104a). (NOT_IMPLEMENTED) (version 25.1.1.1) (from [::ffff:127.0.0.1]:56486) (in query: insert into in_dist from infile 'test.tsv' settings min_insert_block_size_rows=0, min_insert_block_size_bytes=0), Stack trace (when copying this message, always include the lines below):
    2025.01.17 20:32:31.248739 [ 1956 ] {5e4cd57a-7f69-4a7d-bc02-5948f528bebc} <Test> TCPHandler: Logs and exception has been sent. The connection is preserved.
    2025.01.17 20:32:31.248768 [ 1956 ] {} <Error> TCPHandler: Code: 48. DB::Exception: Method write is not supported by storage SystemOne: while pushing to view default.mv (3e957e5d-8df3-4b84-b940-2aa5d4ad104a). (NOT_IMPLEMENTED), Stack trace (when copying this message, always include the lines below):
    2025.01.17 20:32:31.254353 [ 1956 ] {} <Error> TCPHandler: Code: 101. DB::Exception: Unexpected packet Data received from client. (UNEXPECTED_PACKET_FROM_CLIENT), Stack trace (when copying this message, always include the lines below):

The reason for this is that the read_all_data is wrongly set after reading external tables.

But note, that this is a minor thing, since it does not affect subsequent queries, since it will receive this unexpected packet in the same connection for the same query and will close it, so the driver should only re-establish connection (as many does).

azat committed:
Avoid leaving connection in broken state after INSERT finishes with exception

In case of exception during processing INSERT, it is possible to leave
in the state when the will be pending Data packets, and you will get
something like this in logs:

    2025.01.17 20:32:31.228897 [ 1956 ] {5e4cd57a-7f69-4a7d-bc02-5948f528bebc} <Debug> executeQuery: (from [::ffff:127.0.0.1]:56486) insert into in_dist from infile 'test.tsv' settings min_insert_block_size_rows=0, min_insert_block_size_bytes=0 (stage: Complete)
    2025.01.17 20:32:31.248667 [ 1956 ] {5e4cd57a-7f69-4a7d-bc02-5948f528bebc} <Error> executeQuery: Code: 48. DB::Exception: Method write is not supported by storage SystemOne: while pushing to view default.mv (3e957e5d-8df3-4b84-b940-2aa5d4ad104a). (NOT_IMPLEMENTED) (version 25.1.1.1) (from [::ffff:127.0.0.1]:56486) (in query: insert into in_dist from infile 'test.tsv' settings min_insert_block_size_rows=0, min_insert_block_size_bytes=0), Stack trace (when copying this message, always include the lines below):
    2025.01.17 20:32:31.248739 [ 1956 ] {5e4cd57a-7f69-4a7d-bc02-5948f528bebc} <Test> TCPHandler: Logs and exception has been sent. The connection is preserved.
    2025.01.17 20:32:31.248768 [ 1956 ] {} <Error> TCPHandler: Code: 48. DB::Exception: Method write is not supported by storage SystemOne: while pushing to view default.mv (3e957e5d-8df3-4b84-b940-2aa5d4ad104a). (NOT_IMPLEMENTED), Stack trace (when copying this message, always include the lines below):
    2025.01.17 20:32:31.254353 [ 1956 ] {} <Error> TCPHandler: Code: 101. DB::Exception: Unexpected packet Data received from client. (UNEXPECTED_PACKET_FROM_CLIENT), Stack trace (when copying this message, always include the lines below):

The reason for this is that the read_all_data is wrongly set after
reading external tables.

But note, that this is a minor thing, since it does not affect
subsequent queries, since it will receive this unexpected packet in the
same connection for the same query and will close it, so the driver
should only re-establish connection (as many does).

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

robot-clickhouse-ci-1 commented:
<!-- automatic status comment for PR #74740 from azat/ClickHouse:insert-fix-connection-after-exception -->
*This is an automated comment for commit c74a5910c7a62e562fa9baed9625ef2caf99bf3b with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/74740/c74a5910c7a62e562fa9baed9625ef2caf99bf3b/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/9dc69ad65e5ba3eeb3324ba3f1d01308008cd16d/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

azat committed:
Send EndOfStream/Exception before starting skipping the input data

Otherwise the query may hang, simple repo:

    chc --query "select sleep(0.1) from system.numbers settings max_block_size = 1 format Native" | chc --max_execution_time 0.3 --timeout_overflow_mode break --query "insert into null_t format Native"

Fixes: 03221_insert_timeout_overflow_mode
Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>
